### DISCUSSION TITLE
TP 5 - Ejercicio 4 - tiempo_t - ¿Tengo que implementar funciones o estructuras?

### DISCUSSION THREAD

**Discussion by LautaroTummino** [INGCOM-UNRN-P1/cursada-2023#168] [Preguntas y respuestas] [answered]
### Desarrollar una estructura y funciones que permitan realizar las siguientes operaciones con tiempo expresado en horas, minutos y segundos.

- *Crear tiempo_t*
- *Mostrar tiempo_t*
- *Sumar dos tiempo_t*
- *Comparar dos tiempo_t, indicando:*

### Consulta, para crear tiempo_t tengo que hacer una función donde el usuario ingrese las horas minutos y segundos, o creando la estructura ya estaría?

~~~
typedef struct 
{
    int horas;
    int minutos;
    int segundos;
} tiempo_t;
~~~

### Saludos!

**Comment by martinvilu** [comment, marked as answer]
¿Como sería esto?

> Consulta, para crear tiempo_t tengo que hacer una función donde el usuario ingrese las horas minutos y segundos, o creando la estructura ya estaría?

Yo te diría, por qué no ambos.

**Reply by LautaroTummino** [reply]
~~~
void creador_tiempo(tiempo_t creando_tiempo)
{
    printf("Ingrese un numero que representa la hora\n");
    scanf("%d", &creando_tiempo.horas);
    printf("Ingrese un numero que representa los minutos\n");
    scanf("%d", &creando_tiempo.minutos);
    printf("Ingrese un numero que representa los segundos\n");
    scanf("%d", &creando_tiempo.segundos);

    printf("El horario ingresado es: %d:%d:%d\n", creando_tiempo.horas, creando_tiempo.minutos, creando_tiempo.segundos);
}
~~~
~~~
typedef struct 
{
    int horas;
    int minutos;
    int segundos;
} tiempo_t;
~~~

¿Algo así?

**Reply by LautaroTummino** [reply]
Entonces defino la estructura, y luego le pido al usuario que ingrese los valores.

Muchas gracias!

**Reply by martinvilu** [reply]
¿Y como 'sale' el tiempo que pedís?

¿Cuál es el propósito de la función?

**Reply by LautaroTummino** [reply]
Y es un procedimiento, por ende ¿solo es printf y scanf?

Digamos entra una copia a la función, de los datos del main que solo están declarados y printea lo que ingresa el usuario, pero ese valor muere ahí.

El propósito debería ser el de retornar la sctruct completada por el usuario digamos, entonces no tiene que ser void tiene que ser del tipo struct 'tiempo_t' y retornar la struct creando_tiempo, en este caso.

**Reply by martinvilu** [reply]
> Y es un procedimiento,

Si

> por ende ¿solo es printf y scanf?

¿Que hace en argumento? Siendo que usas scanf de entrada.

> Digamos entra una copia a la función, de los datos del main que solo están declarados y printea lo que ingresa el usuario, pero ese valor muere ahí.

Pero llegan como una copia, pero no le das uso; lo sobreescribis de entrada.

> El propósito debería ser el de retornar la sctruct completada por el usuario digamos,

Si, y esto no estaría sucediendo en el código que mostras.

> entonces no tiene que ser void tiene que ser del tipo struct 'tiempo_t' y retornar la struct creando_tiempo, en este caso.

Si, y¿que haces con el argumento?

**Reply by LautaroTummino** [reply]
Si, y¿que haces con el argumento? -> No termino de entender la pregunta.

Pero básicamente el argumento es ¿una variable tipo estructura verdad?, que cuenta con tres miembros 'horas', 'minutos' 'segundos' y es la que mediante scanfs voy a pedirle al usuario que le asigne valor a cada uno.

Dejo un ej a ver si va mejor. 

~~~
typedef struct
{
    int horas;
    int minutos;
    int segundos;
} tiempo_t;


tiempo_t creador_tiempo(tiempo_t creando_tiempo)
{
    printf("Ingrese un numero que representa la hora\n");
    scanf("%d", &creando_tiempo.horas);
    printf("Ingrese un numero que representa los minutos\n");
    scanf("%d", &creando_tiempo.minutos);
    printf("Ingrese un numero que representa los segundos\n");
    scanf("%d", &creando_tiempo.segundos);

    return creando_tiempo;
}

int main()
{

    tiempo_t creando_tiempo;
    creando_tiempo.horas = 0;
    creando_tiempo.minutos = 0;
    creando_tiempo.segundos = 0;

    creando_tiempo = creador_tiempo(creando_tiempo);
    muestro_horario(creando_tiempo);
}
~~~

**Reply by martinvilu** [reply]
> Si, y¿que haces con el argumento? -> No termino de entender la pregunta.

¿Para que esta el argumento?

**Reply by LautaroTummino** [reply]
para crear el tiempo del usuario?

**Reply by martinvilu** [reply]
Estoy insistiendo tanto con este tema, porque lo primero que haces con el argumento es tirarlo por la ventana con el `scanf`

¿Cómo es el comentario de documentación del argumento?

**Reply by LautaroTummino** [reply]
No lo documente todavía, tenia la duda si debíamos hacer una función que el usuario tenga cargar valores, o si directamente se los asignaba en el main. Ahora tengo dudas de hasta mi nombre... no jaja pero bueno intentando pensar con las preguntas.

**Comment by AnGys09** [comment]
Entonces, el usuario estaría dando datos? de  tiempo_t ejemplo , por que la struct la tenemos que hacer antes de la función..... cri cri....

**Reply by LautaroTummino** [reply]
Creo que si, mi duda estaba en si era necesaria esa función, o si simplemente definimos la struct, asignamos valores y hacemos el resto de funciones para ir operando.

Pero hacemos ambas!

**Reply by martinvilu** [reply]
> Entonces, el usuario estaría dando datos? de tiempo_t ejemplo , por que la struct la tenemos que hacer antes de la función..... cri cri....

Hay dos cosas acá, una es la función que 'llena' la `struct`.

La otra, es definir la estructura de la cual después se utiliza para crear variables (y argumentos).

**Reply by martinvilu** [reply]
¿Qué pasó? (que veo un 😕)

**Reply by AnGys09** [reply]
no termino de entender bien el tema de la funcion procedimiento, para  hacer la struct, quiza es mas facil de lo que pienso, pero en este momento tengo un remolino en la capocha

**Reply by martinvilu** [reply]
Un `struct` es un bloque compuesto de tipos de datos sueltos, hora, minuto y segundo como una sola cosa (y en este caso, tres `int`.

El comportamiento de este bloque, es idéntico a cualquier otra variable común, se pasan por copia y se pueden retornar como cualquier otra variable.

Eso, por un lado.

Un procedimiento es una función que no tiene un retorno. `void hace_algo();`, el tipo de retorno de una función de estas características es el "no-tipo", indicado con la palabra reservada `void`

"hacer el `struct`", sería algo como `tiempo_t cargar();`
